File: src/main/java/am/adrian/dungeonkeeper/game/character/Goblin.java
package am.adrian.dungeonkeeper.game.character;

import am.adrian.dungeonkeeper.common.constant.Direction;
import am.adrian.dungeonkeeper.common.coords.MutableCoords;
import am.adrian.dungeonkeeper.common.health.Health;
import am.adrian.dungeonkeeper.common.object.*;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

@RequiredArgsConstructor
@Getter
@Setter
public class Goblin implements Creature, Destroyable, Walks, Swims, Emotional, Levelable {

    private static final String TEXTURE = "goblin.png";

    private static final Logger logger = LogManager.getLogger(Goblin.class);

    private final MutableCoords coords = new MutableCoords();
    private final Health health;
    private final int width;
    private final int height;

    private int level = 1;
    private Mood mood = Mood.HAPPY;

    @Override
    public String getTexture() {
        return TEXTURE;
    }

    @Override
    public int getWidth() {
        return width;
    }

    @Override
    public int getHeight() {
        return height;
    }

    @Override
    public void destroyedAction() {
        logger.debug("{} has been destroyed/killed.", this.getClass().getName());
    }

    @Override
    public void walk(Direction dir) {
        move(dir, walkDelta());
    }

    @Override
    public int walkDelta() {
        return 5;
    }

    @Override
    public void swim(Direction dir) {
        move(dir, swimDelta());
    }

    @Override
    public int swimDelta() {
        return 1;
    }

    @Override
    public void levelUp() {
        ++level;
    }

    public void move(Direction dir, int delta) {
        switch (dir) {
            case UP -> coords.incY(-delta);
            case DOWN -> coords.incY(delta);
            case LEFT -> coords.incX(-delta);
            case RIGHT -> coords.incX(delta);
            default -> throw new RuntimeException("Unknown direction");
        }
    }

    @Override
    public boolean collides(@NotNull GameObject gameObject) {
        return false;
    }
}

File: src/main/kotlin/am/adrian/dungeonkeeper/model/gameobject/creature/Goblin.kt
package am.adrian.dungeonkeeper.model.gameobject.creature

import am.adrian.dungeonkeeper.behavior.character.GoblinBehaviorTree
import am.adrian.dungeonkeeper.constant.Direction
import am.adrian.dungeonkeeper.helper.Players
import am.adrian.dungeonkeeper.model.coords.MutableCoords
import am.adrian.dungeonkeeper.model.gameobject.GameObject
import am.adrian.dungeonkeeper.model.gameobject.trait.*
import am.adrian.dungeonkeeper.model.gameobject.trait.HasMood.Mood
import am.adrian.dungeonkeeper.model.health.Health
import am.adrian.dungeonkeeper.model.player.Player
import kotlin.math.max
import kotlin.math.min

private const val TEXTURE = "goblin.png"

class Goblin(
    private val width: Int,
    private val height: Int,
    private val health: Health,
    private var hungerLevel: Int = 100,
    private val coords: MutableCoords = MutableCoords(),
    private var belongsTo: Player = Players.NEUTRAL_PLAYER,
    private var mood: Mood = Mood.HAPPY,
    private var level: Int = 1,
) : Creature, HasHealth, HasMood, HasBehavior, CanEat, CanWalk, CanSwim, CanLevel, CanFear {

    override fun collides(gameObject: GameObject): Boolean = false

    override fun getCoords(): MutableCoords = coords

    override fun getTexture(): String = TEXTURE

    override fun getLevel(): Int = level

    override fun levelUp() {
        ++level
    }

    override fun getWidth(): Int = width

    override fun getHeight(): Int = height

    override fun belongsTo(): Player = belongsTo

    override fun walk(dir: Direction) = move(dir, walkDelta())

    override fun walkDelta(): Int = 5

    override fun swim(dir: Direction) = move(dir, swimDelta())

    override fun swimDelta(): Int = 1

    override fun getFearThreshold(): Int = getLevel() * 2

    override fun move(dir: Direction, delta: Int) {
        when (dir) {
            Direction.UP -> coords.incY(-delta)
            Direction.DOWN -> coords.incY(delta)
            Direction.LEFT -> coords.incX(-delta)
            Direction.RIGHT -> coords.incX(delta)
        }
    }

    override fun getHealth(): Health = health

    override fun getMood(): Mood = mood

    override fun setMood(mood: Mood) {
        this.mood = mood
    }

    override fun hungerLevel(): Int = hungerLevel

    override fun adjustHunger(level: Int) {
        hungerLevel = min(CanEat.MAX_HUNGER_LEVEL, max(0, hungerLevel + level))
    }

    override fun isHungry(): Boolean = hungerLevel < 20

    override fun getBehaviorTree() = GoblinBehaviorTree(this)
}
